### PR TITLE
fix: api/enable ts strict

### PR DESCRIPTION
**Pull request by iartemiev**
#### Description of changes

Enables TS strict mode and `noImplicitAny` in `api-graphql`. 
Note there are still some _explicit_ `any` in the source, but `tsc` is ok with that. 

Co-authored by @erinleigh90 and @HuiSF 

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/callable-release-verification.yml [hunk 1]
```diff
@@ -14,17 +14,17 @@ jobs:
...
     secrets: inherit
     uses: ./.github/workflows/callable-prebuild-samples-staging.yml
-  e2e:
-    needs:
-      - prebuild-macos
-      - prebuild-ubuntu
-      - prebuild-samples-staging
-    secrets: inherit
-    uses: ./.github/workflows/callable-e2e-tests.yml
+  # e2e:
```
Comment (iartemiev) [inline .github/workflows/callable-release-verification.yml:17]: Unintended change will revert.

File: packages/aws-amplify/__tests__/initSingleton.test.ts [hunk 2]
```diff
@@ -66,6 +66,13 @@ describe('initSingleton (DefaultAmplify)', () => {
 		mockCognitoUserPoolsTokenProviderSetKeyValueStorage.mockReset();
 		mockAmplifySingletonConfigure.mockReset();
 		mockAmplifySingletonGetConfig.mockReset();
+
+		mockAmplifySingletonConfigure.mockImplementation((_, libraryOptions) => {
+			AmplifySingleton.libraryOptions =
+				libraryOptions ?? AmplifySingleton.libraryOptions;
+		});
+		// reset to its initial state
+		AmplifySingleton.libraryOptions = {};
```
Comment (iartemiev) [inline packages/aws-amplify/__tests__/initSingleton.test.ts:75]: @erinleigh90 - do you know if this is intentional or just a bad merge?

File: packages/api/tsconfig.json [hunk 3]
```diff
@@ -2,8 +2,8 @@
 	"extends": "../tsconfig.base.json",
 	"compilerOptions": {
 		"importHelpers": true,
-		"strict": false,
-		"noImplicitAny": false,
+		"strict": true,
+		"noImplicitAny": true,
```
Comment (jimblanc) [inline packages/api/tsconfig.json:6]: ❤️

**Review by jimblanc — DISMISSED** [review]
LGTM, pending what Erin says about the `initSingleton.test.ts` change

File: packages/api-graphql/package.json [hunk 4]
```diff
@@ -98,16 +99,11 @@
 	"jest": {
 		"globals": {
 			"ts-jest": {
-				"diagnostics": {
-					"pathRegex": "(/__tests__/.*|\\.(test|spec))\\.(tsx?|jsx?)$"
-				},
+				"diagnostics": false,
```
Comment (iartemiev) [inline packages/api-graphql/package.json:102]: Disabling `diagnostics` as it was giving us false errors, such as 
> Could not find a declaration file for module 'rxjs'

Even though `rxjs` types are definitely present in node_modules and `tsc` is happy with the types it emits. Other categories have `diagnostics` disabled as well FWIW

File: packages/aws-amplify/__tests__/initSingleton.test.ts (hunk 2 above)
Comment (erinleigh90) [reply packages/aws-amplify/__tests__/initSingleton.test.ts:75]: Bad merge, I was just going to ask you about this as well 😂

Comment (iartemiev) [reply packages/aws-amplify/__tests__/initSingleton.test.ts:75]: Cool! I'll revert

Comment (iartemiev) [reply packages/aws-amplify/__tests__/initSingleton.test.ts]: Done

**Review by jimblanc — DISMISSED** [review]

File: packages/api-graphql/src/types/index.ts [hunk 5]
```diff
@@ -364,6 +364,16 @@ export const __amplify = Symbol('amplify');
 export const __authMode = Symbol('authMode');
 export const __authToken = Symbol('authToken');
 
+export type ClientWithoutModels =
+	| V6Client<never>
+	| V6ClientSSRRequest<never>
+	| V6ClientSSRCookies<never>;
+
```
Comment (erinleigh90) [inline packages/api-graphql/src/types/index.ts:371]: I don't think this type ended up being used

**Review by erinleigh90 — APPROVED** [review]
LGTM, but I think we can remove the type ClientWithoutModel

**Review by jimblanc — APPROVED** [review]

File: packages/api-graphql/src/types/index.ts (hunk 5 above)
Comment (iartemiev) [reply packages/api-graphql/src/types/index.ts]: Done

**Review by erinleigh90 — DISMISSED** [review]

File: packages/api-graphql/package.json (hunk 4 above)
Comment (svidgen) [reply packages/api-graphql/package.json:102]: That's a little terrifying. Is there a path to re-enabling diagnostics?

**Review by svidgen — APPROVED** [review]
No obvious blunders, but did we sanity check with a sample app that the types on `graphql()` using branded queries still work as expected?

**Review by david-mcafee — APPROVED** [review]

**Review by jimblanc — APPROVED** [review]

**Comment by iartemiev** [comment]
> No obvious blunders, but did we sanity check with a sample app that the types on `graphql()` using branded queries still work as expected?

@svidgen - validated manually in the IDE using the `v6-test.ts` test

**Review by svidgen — APPROVED** [review]

**Pull request merged**